sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -940,7 +948,7 @@ object SparkSession {
         }
 
         session = new SparkSession(sparkContext, None, None, extensions)
-        options.foreach { case (k, v) => session.sessionState.conf.setConfString(k, v) }
+        options.foreach { case (k, v) => session.initialSessionOptions.put(k, v) }
```

Comment:
I tried to fix this problem in https://github.com/apache/spark/pull/18172/files#diff-d91c284798f1c98bf03a31855e26d71cL938 , maybe you can pick it up? Your approach adds a new concept(`initialSessionOptions`), while `SparkSession` initialization is already very complex...